### ISSUE TITLE
TERM/INT signals do not propagate to children

### ISSUE THREAD

**Issue by proger** [MichaelXavier/Angel#40]
Hi,

I think I found a severe issue in Angel:

Steps to reproduce:

1. run angel with a single arbitrary process
2. send a parent angel either TERM or INT signal (^C won't work because it sends INT to the whole *process group*)
3. watch the parent orphan its children

```
% pstree -p $(pgrep angel | head -1)
-+= 00001 root /sbin/launchd
 \-+= 00341 vladki /Applications/iTerm.app/Contents/MacOS/iTerm -psn_0_40970
   \-+= 00453 root login -fp vladki
     \-+= 00456 vladki -zsh
       \-+= 03508 vladki dist/build/angel/angel -v2 stubborn.conf
         \-+- 03509 vladki dist/build/angel/angel -v2 stubborn.conf
           \--- 03510 vladki test/test_jobs/StubbornJob
% kill -TERM 3508
% pstree -p $(pgrep angel | head -1)
-+= 00001 root /sbin/launchd
 \-+- 03509 vladki dist/build/angel/angel -v2 stubborn.conf
   \--- 03510 vladki test/test_jobs/StubbornJob
```

Extended logs:

```
[2015/05/21 14:48:43] {main} INT | TERM received; initiating shutdown...
[2015/05/21 14:48:43] {main}   1. Clearing config
[2015/05/21 14:48:43] {main}   2. Forcing sync to kill running processes
[2015/05/21 14:48:43] {process-monitor} Must kill=0, must start=0
[2015/05/21 14:48:43] {main} That's all folks!
```

Anticipated cause:

Angel tries to [modify](https://github.com/MichaelXavier/Angel/blob/master/src/Angel/Job.hs#L137) a `TVar` [inside `forkProcess`](https://github.com/MichaelXavier/Angel/blob/master/src/Angel/Job.hs#L155-L171) that obviously doesn't modify a parent process' `TVar`.

A current workaround is to simply kill the whole process group of Angel, but this is definitely not the cleanest solution.

**Comment by proger** [comment]
The same issue applies when you modify the config and *remove* processes from it (they never get stopped)

**Comment by ignesco** [comment]
I have not had much time to look into it, but I have managed to replicate the issues that you have decribed.  I will try to give some initial feedback this weekend.

**Comment by ignesco** [comment]
So there are three things that appear to be contributing to this issue.

1. [onValidHandleAction](https://github.com/MichaelXavier/Angel/blob/master/src/Angel/Job.hs#L117-L120) is trying to register the process handle(s) from the supervisee, this call is failing to register in the main configuration because it is called via [a new process](https://github.com/MichaelXavier/Angel/blob/master/src/Angel/Job.hs#L169-L172) it should be the process id of the [forked process](https://github.com/MichaelXavier/Angel/blob/master/src/Angel/Job.hs#L155) that is actually registered as this is the controlling process for the supervisee
2. the forked process does not correctly terminate the supervisee (and the log process is there is one) it should take responsibility rather just *trying* to store them in the current configuration
3. the forked process does not respond to TERM and INT.  The new forked process needs to have handlers [like in Main.hs](https://github.com/MichaelXavier/Angel/blob/master/src/Angel/Main.hs#L107-L108), so that it can handle the proper termination of the supervisee as in point **2.**

All these points should addess the HUP configuration re-read problem as well.

@proger I did not find this whilst testing because I was relying on Crtl-C and did not realise that it appeared to be working because of the termination of the Process group.

@MichaelXavier what is the best way to proceed how about a new issue branch so that I can send a pull request and then get feedback from @proger

It will take me a few days due to other work commitments, but I have a good handle on a fix.

@proger Thanks for taking the time to report this issue.

**Comment by MichaelXavier** [comment]
@ignesco Thanks for the writeup. Sounds like you have a plan. I'd definitely accept this as a PR. That's how I've been doing it as well, letting the person who files the issue test it out and giving it a week or so for feedback before merge. The only thing I'll add is that if you can demonstrate this in the test suite, having a test for this would be great. Let me know if you have any further questions, and thanks again for taking on the work!

**Comment by ghost** [comment]
@MichaelXavier I booked the day off today to finialise the work for this issue and I coded up a solution in about an hour or two, and it worked for the simple test case, but when I stress tested it, it has hilighted a much more complicated and tricky side effects.  I have spend most of the day trying to get to the bottom of what the problem is but to  no avail.  When count is greater than 1 for particular configuration, it has the very bad effect that the `forkProcess` function does not actually fire for any more that a couple of instances in my test of count = 10, at the moment I am tempted to say it is GHC and forkProcess not playing with multiple processors (as per the warning in its hoogle page) or just plain old lazy evaluation but as to which one, I'm not sure.

So the short of it is that I do not think my fix will work for anything but the simplest case.

At this stage for the sake of stability, I would recommend backing out the "run as user" change until I can come up with a more robust solution to forkProcess not working, what do you think?  I don't mind do the backing out too, having proposed and implemented it in the first place I feel that I should clean up too.

Can you let me know what you would like me to do?

**Comment by proger** [comment]
@ignesco thanks for putting your effort into it!

As a workaround, you could also wrap the command into `sudo -u` or something like that while you're looking for the nicer fix.

**Comment by MichaelXavier** [comment]
@ignesco Sorry about the delay. You caught me right at the beginning of a vacation. I've given you commit access to the repo, since it would seem you use it in production more than I do. I'd say in general to use your best judgement with the following values in mind:

1. Stability is important. We should avoid breaking changes.
2. We should be aiming for the most robust production behavior we can. I'd rather avoid pushing a feature that we're iffy on. By all means, if there's a question as to what angel should do in a particular instance, we should be guided by the principle of least surprise and if possible, copy the behavior of the other venerable process monitors.

For this particular instance, sounds like backing out the change is the right thing to do until you're really confident in it. I'm sure you'll use good judgment here. Thanks!

**Comment by ignesco** [comment]
@MichaelXavier I have pushed an issue branch for the initial fix for this issue (40-term-int-propogation-01).  I opted for the the interim suggestion of @proger of "sudo -u" do you think this would this be OK?  Also the test for "forcefully kills stubborn processes" seems to fail on my machine, but it was failing in master SHA1-ID: d5e3e199223 would you still be OK with this test failing?  Also if there anything I need to do with Travis to get it to run tests for this branch?

@proger I got side tracked with trying to get to the bottom of the issues that I saw whilst trying to fix this issue.  To get things moving again, I have coded the removal of forProcess for now, with your suggestion of "sudo -u".  The changes are in the branch (40-term-int-propogation-01).  If you get a chance can you try these latest changes to confirm that it fixes the original issue that you were experiencing?

On a side-note with the issue that I saw whilst trying to fix this issue, I think I have narrowed it down to something to do with the child processing seein stdout as being locked so it just waits ad infinitum and cannot respond to any of the parent's signals.  I have some other things to try around this, so I'll continue with trying to get it to work.

**Comment by MichaelXavier** [comment]
@ignesco Seems like the tests actually ran and passed for everything except for 7.10.1: https://travis-ci.org/MichaelXavier/Angel

Its probably a travis file issue. I'm not too worried about it. I say that between you and @proger you have a pretty good idea of the issue at hand so once you get it solved to your satisfaction, I'll be happy to merge it. Using `sudo` still feels weird to me. The possible issues are either env issues where Angel won't be able to find sudo or possibly systems where sudo doesn't exist. Does Mac have sudo?

**Comment by proger** [comment]
@MichaelXavier by suggesting sudo I mean it should not be a feature of the app per se, a user can just use sudo whenever one feels comfortable (we're chpst from runit, for example).

**Comment by ghost** [comment]
@MichaelXavier and @proger , I have tried lots of different options for fixing this issue, non of which have actually resulted in anthing that I could have confidence in.  After considering other options to get this issue resolved, the two steps that I think are now most viable are:
1) Remove the forkProcess function as I consider this to be the root cause
2) And either...
     a) Choose a more global approach and use setuid to set the user id across the whole process on initial start up, reading the user from the command line options or from the configutation file.
     b) Just remove the "run as a different user" functionality

So what does everyone think?  @proger you said your system deployment uses chpst, would it be beneficial to you to be able to do this directly using just Angel or would you just stick with chpst anyway?  Personally I would like this functionality to be embeded in the Angel config file because backing the config file up would be backing up more of the overall system deployment configuation.

Thanks for waiting on this issue, new baby took precedence.

**Comment by MichaelXavier** [comment]
@ignesco I think you've got the right idea here. Having the config file be the one source of truth seems to be a good goal. As for which of a/b to choose, I'll defer to you guys. I'm not currently using this project in production and value the opinion of people using it in earnest most. I don't have too much time to devote to this project now so I'm happy to let you guys decide what is best and support where I can.

**Comment by proger** [comment]
You may be interested in using https://github.com/haskell/process/pull/45 instead of `forkProcess`

**Comment by ghost** [comment]
Hi All,
Sorry I have been off grid for a while. @proger the https://github.com/haskell/process/pull/45 still has all the issues with the locking stdout problem.  My plan is to finish off the implementation leaving in the command line argument option and implementing the suggested configuration file option. This way anyone using the current version does not have a nasty surprise. Hopefully, if there are no objections we should be able to close this issue very soon.

Thanks for all your patience.

**Comment by ignesco** [comment]
Hi All,
I have pushed a potential fix for the original issue, whilst keeping the flexability for the user to change the executing user, with both -u and user= options. the can be found here
https://github.com/MichaelXavier/Angel/tree/40-term-int-propogation-02
@proger would it be possible to retest this issue with this branch?  @MichaelXavier and @proger please let me know your thoughts and hopefully we can get closer to fixing this issue.

**Comment by ignesco** [comment]
Hi All,
Is there someone that could review this fix before I push it to the master branch.  Would be nice to get this in as it has been an issue for some time now.  I was hoping to push my this coming Sunday 2016-09-04  Many thanks.

**Comment by MichaelXavier** [comment]
@ignesco I don't really have much time for this project and I'm not actively using it right now. The build failure looks trivial though so I wouldn't worry about. This project could probably be set up to use stack so different GHC versions can be used easily.

BTW if any of you are actively relying on Angel and would like commit bit let me know and I'll make it so.

**Comment by ignesco** [comment]
Thanks for everyone's patience with this one - this issue is fixed with commit SHA1-ID 85291481993ca4f34310c1b9e74f26bf565dac0f
